syntax = "proto3";

package plosolver.auth;

option go_package = "github.com/plosolver/backend/protos/auth";
option java_multiple_files = true;
option java_package = "com.plosolver.grpc.auth";
option java_outer_classname = "AuthProto";

import "google/protobuf/timestamp.proto";
import "common.proto";

// =============================================================================
// Authentication Service
// =============================================================================

service AuthService {
  rpc Register(RegisterRequest) returns (AuthResponse);
  rpc Login(LoginRequest) returns (AuthResponse);
  rpc Logout(LogoutRequest) returns (LogoutResponse);
  rpc RefreshToken(RefreshTokenRequest) returns (AuthResponse);
  rpc VerifyToken(VerifyTokenRequest) returns (VerifyTokenResponse);
  rpc ForgotPassword(ForgotPasswordRequest) returns (ForgotPasswordResponse);
  rpc ResetPassword(ResetPasswordRequest) returns (ResetPasswordResponse);
  rpc ChangePassword(ChangePasswordRequest) returns (ChangePasswordResponse);
  rpc GetProfile(GetProfileRequest) returns (GetProfileResponse);
  rpc UpdateProfile(UpdateProfileRequest) returns (UpdateProfileResponse);
  rpc Waitlist(WaitlistRequest) returns (WaitlistResponse);
  rpc GoogleAuth(GoogleAuthRequest) returns (AuthResponse);
}

message RegisterRequest {
  string email = 1;
  string password = 2;
  string username = 3;
  string invite_code = 4;
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message LogoutRequest {
  string refresh_token = 1;
}

message LogoutResponse {
  bool success = 1;
  string message = 2;
}

message RefreshTokenRequest {
  string refresh_token = 1;
}

message VerifyTokenRequest {
  string token = 1;
}

message VerifyTokenResponse {
  bool valid = 1;
  plosolver.common.User user = 2;
  plosolver.common.Error error = 3;
}

message ForgotPasswordRequest {
  string email = 1;
}

message ForgotPasswordResponse {
  bool success = 1;
  string message = 2;
}

message ResetPasswordRequest {
  string token = 1;
  string new_password = 2;
}

message ResetPasswordResponse {
  bool success = 1;
  string message = 2;
}

message ChangePasswordRequest {
  string current_password = 1;
  string new_password = 2;
}

message ChangePasswordResponse {
  bool success = 1;
  string message = 2;
}

message GetProfileRequest {
  string user_id = 1;
}

message GetProfileResponse {
  plosolver.common.User user = 1;
  plosolver.common.Error error = 2;
}

message UpdateProfileRequest {
  string user_id = 1;
  string username = 2;
  string email = 3;
}

message UpdateProfileResponse {
  plosolver.common.User user = 1;
  plosolver.common.Error error = 2;
}

message WaitlistRequest {
  string email = 1;
  string name = 2;
  string company = 3;
  string role = 4;
}

message WaitlistResponse {
  bool success = 1;
  string message = 2;
}

message GoogleAuthRequest {
  string id_token = 1;
}

message AuthResponse {
  plosolver.common.User user = 1;
  string access_token = 2;
  string refresh_token = 3;
  string token_type = 4;
  int32 expires_in = 5;
  plosolver.common.Error error = 6;
}
